app/services/fact_importer.rb

```diff
@@ -66,11 +68,17 @@ def add_new_facts
     facts_to_create = facts.keys - db_facts.keys
     # if the host does not exists yet, we don't have an host_id to use the fact_values table.
     if facts_to_create.present?
-      method          = host.new_record? ? :build : :create!
-      fact_names      = fact_name_class.group(:name).maximum(:id)
+      method = host.new_record? ? :build : :create!
+      # :type is needed because custom facts usually inherits from FactName so they would be included in the list
+      fact_names = fact_name_class.where(:type => fact_name_class).group(:name).maximum(:id)
       facts_to_create.each do |name|
-        host.fact_values.send(method, :value => facts[name],
-                              :fact_name_id  => fact_names[name] || fact_name_class.create!(:name => name).id)
+        begin
+          host.fact_values.send(method, :value => facts[name],
+                                :fact_name_id  => fact_names[name] || fact_name_class.create!(:name => name).id)
```

Comment:
I haven't realized that the `fact_name_class` could actually be a parent (expected all the actually used fact types would be just children of some abstract parent). Makes sense now. Thanks